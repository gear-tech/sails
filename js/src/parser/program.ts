import { FuncParam, Service } from './service.js';
import { Type, WithDef } from './types.js';
import { getName } from './util.js';
import { Base } from './visitor.js';

export class Program {
  private _services: Service[];
  private _types: Map<number, Type>;
  private _context: Map<number, WithDef>;
  private _ctor: Ctor;

  constructor() {
    this._services = [];
    this._types = new Map();
    this._context = new Map();
  }

  addService(service: Service) {
    this._services.push(service);
  }

  addType(type: Type) {
    const id = type.rawPtr;
    this._types.set(id, type);
    this._context.set(id, type);
    return id;
  }

  get services(): Service[] {
    return this._services;
  }

  get ctor(): Ctor {
    return this._ctor;
  }

  getType(id: number): Type {
    return this._types.get(id);
  }

  getContext(id: number): any {
    return this._context.get(id);
  }

  addContext(id: number, ctx: any) {
    this._context.set(id, ctx);
  }

  get types(): Type[] {
    return Array.from(this._types.values());
  }

  getTypeByName(name: string): Type {
    const types = this.types.filter((type) => type.name === name);
    if (types.length > 1) throw new Error(`multiple types found with name ${name}`);
    if (types.length === 0) throw new Error(`no type found with name ${name}`);

    return types[0];
  }

  addCtor(ctor: Ctor) {
    this._ctor = ctor;
  }
}

export class Ctor extends Base {
  public readonly funcs: CtorFunc[];

  constructor(ptr: number, memory: WebAssembly.Memory) {
    super(ptr, memory);

    this.funcs = [];
  }

  addFunc(func: CtorFunc) {
    this.funcs.push(func);
  }
}

export class CtorFunc extends Base {
  private _params: Map<number, FuncParam>;
  public readonly name: string;

  constructor(ptr: number, memory: WebAssembly.Memory) {
    super(ptr, memory);

    const { name, offset } = getName(ptr, this.offset, memory);
    this.name = name;
    this.offset = offset;

    this._params = new Map();
  }

  addFuncParam(ptr: number, param: FuncParam) {
    this._params.set(ptr, param);
  }

  get params(): FuncParam[] {
    if (this._params.size === 0) return [];

    return Array.from(this._params.values());
  }
}
