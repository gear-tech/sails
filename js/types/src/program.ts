import { ISailsDocs } from './idl';
import { ISailsEnumVariant, ISailsType, ISailsTypeDef } from './types';

export interface ISailsProgram {
  readonly services: ISailsService[];
  readonly ctor: ISailsCtor;
  readonly types: ISailsType[];

  getTypeByName(name: string): ISailsType;
  getType(id: number): ISailsType;
}

export interface ISailsService {
  readonly funcs: ISailsServiceFunc[];
  readonly events: ISailsServiceEvent[];
  readonly name: string;
}

export interface ISailsServiceFunc extends ISailsDocs {
  readonly name: string;
  readonly isQuery: boolean;
  readonly params: ISailsFuncParam[];
  readonly def: ISailsTypeDef;
}

export interface ISailsCtor {
  readonly funcs: ISailsCtorFunc[];
}

export interface ISailsCtorFunc extends ISailsDocs {
  readonly name: string;
  readonly params: ISailsFuncParam[];
}

export interface ISailsFuncParam {
  readonly name: string;
  readonly def: ISailsTypeDef;
}

export type ISailsServiceEvent = ISailsEnumVariant;
