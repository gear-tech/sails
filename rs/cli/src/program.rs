use anyhow::Result;
use cargo_generate::{GenerateArgs, TemplatePath};
use std::{env, path::PathBuf};

const TEMPLATES_BRANCH_ENV_VAR: &str = "SAILS_CLI_TEMPLATES_BRANCH";
const TEMPLATES_REPO: &str = "https://github.com/gear-tech/sails.git";
const PROGRAM_TEMPLATE_PATH: &str = "templates/program";

pub struct ProgramGenerator {
    path: String,
    name: Option<String>,
    with_client: bool,
    with_gtest: bool,
}

impl ProgramGenerator {
    pub fn new(path: String) -> Self {
        Self {
            path,
            name: None,
            with_client: false,
            with_gtest: false,
        }
    }

    pub fn with_name(self, name: Option<String>) -> Self {
        Self { name, ..self }
    }

    pub fn with_client(self, with_client: bool) -> Self {
        Self {
            with_client,
            ..self
        }
    }

    pub fn with_gtest(self, with_gtest: bool) -> Self {
        Self {
            with_gtest,
            with_client: self.with_client | with_gtest,
            ..self
        }
    }

    pub fn generate(self) -> Result<()> {
        let template_path = TemplatePath {
            auto_path: Some(PROGRAM_TEMPLATE_PATH.into()),
            git: Some(TEMPLATES_REPO.into()),
            branch: env::var(TEMPLATES_BRANCH_ENV_VAR).ok(),
            ..TemplatePath::default()
        };

        let (path, name) = if self.name.is_none() {
            let path_buf = PathBuf::from(&self.path);
            let path = path_buf.parent().map(|p| p.to_path_buf());
            let name = path_buf.file_name().map(|n| {
                n.to_str()
                    .expect("unreachable as was built from UTF-8")
                    .to_string()
            });
            (path, name)
        } else {
            (Some(PathBuf::from(&self.path)), self.name)
        };

        let generate_args = GenerateArgs {
            template_path,
            name,
            destination: path,
            silent: true,
            define: vec![
                format!("with-client={}", self.with_client),
                format!("with-gtest={}", self.with_gtest),
            ],
            ..GenerateArgs::default()
        };
        cargo_generate::generate(generate_args)?;
        Ok(())
    }
}
