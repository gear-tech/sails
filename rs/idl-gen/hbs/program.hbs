program {{name}} {
    constructors {
        {{#each ctors}}
        {{#each docs}}
        /// {{{this}}}
        {{/each}}
        {{name}}({{#each args}}{{#if @index}}, {{/if}}{{name}}: {{{lookup @root.program.concrete_names type}}}{{/each}});
        {{#unless @last}}

        {{/unless}}
        {{/each}}
    }
    {{#if services}}

    services {
        {{#each services}}
        {{this}},
        {{/each}}
    }
    {{/if}}
    {{#if types}}

    types {
{{#each types}}
{{#if this.enum}}
        enum {{this.enum.own_name}} {
        {{#each this.enum.fields}}
            {{own_name}}{{~#if fields}}
            {{~#if fields.[0].name}} {
                {{#each fields}}
                {{name}}: {{{type_name}}},
                {{/each}}
            }
            {{~else~}}
            ({{#each fields}}{{{type_name}}}{{#unless @last}}, {{/unless}}{{/each}})
            {{~/if}}{{~/if}},
        {{/each}}
        }
{{else}}
    {{#if this.struct.fields}}
        {{#if this.struct.fields.[0].name}}
        struct {{this.struct.own_name}} {
            {{#each this.struct.fields}}
            {{name}}: {{{type_name}}},
            {{/each}}
        }
        {{else}}
        struct {{this.struct.own_name}}({{#each this.struct.fields}}{{{type_name}}}{{#unless @last}}, {{/unless}}{{/each}});
        {{/if}}
    {{else}}
        struct {{this.struct.own_name}};
    {{/if}}
{{/if}}
{{#unless @last}}

{{/unless}}
{{/each}}
    }
    {{/if}}
}
