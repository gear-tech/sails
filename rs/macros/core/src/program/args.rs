use crate::sails_paths;
use proc_macro2::Span;
use proc_macro_error::abort;
use std::collections::BTreeSet;
use syn::{
    parse::{Parse, ParseStream},
    punctuated::Punctuated,
    Path, Token,
};

#[derive(Debug, Default, PartialEq)]
pub(super) struct ProgramArgs {
    handle_reply: Option<Path>,
    handle_signal: Option<Path>,
    sails_path: Option<Path>,
}

impl ProgramArgs {
    pub fn handle_reply(&self) -> &Option<Path> {
        &self.handle_reply
    }

    pub fn handle_signal(&self) -> &Option<Path> {
        &self.handle_signal
    }

    pub fn sails_path(&self) -> syn::Path {
        sails_paths::sails_path_or_default(self.sails_path.clone())
    }
}

impl Parse for ProgramArgs {
    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
        let punctuated: Punctuated<ProgramArg, Token![,]> = Punctuated::parse_terminated(input)?;
        let mut attrs = ProgramArgs {
            handle_reply: None,
            handle_signal: None,
            sails_path: None,
        };
        let mut existing_attrs = BTreeSet::new();

        for ProgramArg {
            name, path, span, ..
        } in punctuated
        {
            let name = name.get_ident().unwrap().to_string();
            if existing_attrs.contains(&name) {
                abort!(span, "parameter already defined");
            }

            match &*name {
                "handle_reply" => {
                    attrs.handle_reply = Some(path);
                }
                "handle_signal" => {
                    attrs.handle_signal = Some(path);
                }
                "crate" => {
                    attrs.sails_path = Some(path);
                }
                _ => abort!(
                    span,
                    "`program` attribute can only contain `handle_reply`, `handle_signal` and `crate` parameters",
                ),
            }

            existing_attrs.insert(name);
        }

        Ok(attrs)
    }
}

struct ProgramArg {
    name: Path,
    path: Path,
    span: Span,
}

impl Parse for ProgramArg {
    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
        let span = input.span();
        let name: Path = input.parse()?;
        let _: Token![=] = input.parse()?;
        let path: Path = input.parse()?;

        Ok(Self { name, path, span })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use proc_macro2::Span;
    use quote::quote;
    use syn::{Ident, PathSegment};

    #[test]
    fn gprogram_parse_attrs() {
        // arrange
        let input = quote!(
            handle_reply = my_handle_reply,
            handle_signal = my_handle_signal,
        );
        let expected = ProgramArgs {
            handle_reply: Some(
                PathSegment::from(Ident::new("my_handle_reply", Span::call_site())).into(),
            ),
            handle_signal: Some(
                PathSegment::from(Ident::new("my_handle_signal", Span::call_site())).into(),
            ),
            sails_path: None,
        };

        // act
        let args = syn::parse2::<ProgramArgs>(input).unwrap();

        // arrange
        assert_eq!(expected, args);
    }

    #[test]
    fn gprogram_parse_crate() {
        // arrange
        let input = quote!(crate = sails_rename,);
        let expected = ProgramArgs {
            handle_reply: None,
            handle_signal: None,
            sails_path: Some(
                PathSegment::from(Ident::new("sails_rename", Span::call_site())).into(),
            ),
        };

        // act
        let args = syn::parse2::<ProgramArgs>(input).unwrap();

        // arrange
        assert_eq!(expected, args);
    }
}
