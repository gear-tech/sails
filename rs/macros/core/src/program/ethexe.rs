use super::*;
use crate::shared;
use convert_case::{Boundary, Case, Casing};
use parity_scale_codec::Encode;
use proc_macro2::TokenStream;
use quote::quote;

/// Generates code
/// ```rust
/// impl sails_rs::solidity::ProgramSignature for MyProgram {
///     const CTORS: &'static [sails_rs::solidity::MethodRoute] = &[(
///         sails_rs::concatcp!(
///             "default", << (u128) as sails_rs::alloy_sol_types::SolValue > ::SolType
///             as sails_rs::alloy_sol_types::SolType > ::SOL_NAME,
///         ),
///         &[28u8, 68u8, 101u8, 102u8, 97u8, 117u8, 108u8, 116u8] as &[u8],
///     ),
/// ];
/// const SERVICES: &'static [(
///     &'static str,
///     &'static [u8],
///     &'static [sails_rs::solidity::MethodRoute],
/// )] = &[
///     (
///         "service",
///         &[28u8, 83u8, 101u8, 114u8, 118u8, 105u8, 99u8, 101u8] as &[u8],
///         <MyService as sails_rs::solidity::ServiceSignature>::METHODS,
///     ),
/// ];
/// const METHODS_LEN: usize = <MyService as sails_rs::solidity::ServiceSignature>::METHODS
///     .len();
/// }
/// ```
pub fn program_signature_impl(program_impl: &ItemImpl, sails_path: &Path) -> TokenStream {
    let (program_type_path, _, _) = shared::impl_type_refs(program_impl.self_ty.as_ref());
    let (generics, program_type_constraints) = shared::impl_constraints(program_impl);

    let program_ctors = discover_program_ctors(program_impl);
    let program_ctor_sigs = program_ctors
        .iter()
        .map(|(handler_route, (handler_fn, _, _))| {
            shared::ethexe::handler_signature(handler_route, handler_fn, sails_path)
        });

    let services_ctors = discover_services_ctors(program_impl);
    let service_signatures =
        services_ctors
            .iter()
            .map(|(service_route, (service_fn, _, unwrap_result))| {
                service_signature(service_route, service_fn, *unwrap_result, sails_path)
            });

    let methods_len_iter = services_ctors
        .iter()
        .map(|(_, (service_fn, _, unwrap_result))| {
            let service_type = shared::unwrap_result_type(&service_fn.sig, *unwrap_result);
            quote!(<#service_type as #sails_path::solidity::ServiceSignature>::METHODS.len())
        });
    let methods_len = if services_ctors.is_empty() {
        quote! {0}
    } else {
        quote! {#(#methods_len_iter) + *}
    };
    quote! {
        impl #generics #sails_path::solidity::ProgramSignature for #program_type_path #program_type_constraints {
            const CTORS: &'static [#sails_path::solidity::MethodRoute] = &[
                #( #program_ctor_sigs )*
            ];
            const SERVICES: &'static [(&'static str, &'static [u8], &'static [#sails_path::solidity::MethodRoute])] = &[
                #( #service_signatures )*
            ];
            const METHODS_LEN: usize = #methods_len;
        }
    }
}

pub fn program_const(program_impl: &ItemImpl, sails_path: &Path) -> TokenStream {
    let (program_type_path, _, _) = shared::impl_type_refs(program_impl.self_ty.as_ref());
    quote! {
        const __CTOR_SIGS: [[u8; 4]; <#program_type_path as #sails_path::solidity::ProgramSignature>::CTORS.len()]
            = #sails_path::solidity::ConstProgramMeta::<#program_type_path>::ctor_sigs();
        const __METHOD_SIGS: [[u8; 4]; <#program_type_path as #sails_path::solidity::ProgramSignature>::METHODS_LEN]
            = #sails_path::solidity::ConstProgramMeta::<#program_type_path>::method_sigs();
        const __METHOD_ROUTES: [(&'static [u8], &'static [u8]); <#program_type_path as #sails_path::solidity::ProgramSignature>::METHODS_LEN]
            = #sails_path::solidity::ConstProgramMeta::<#program_type_path>::method_routes();
    }
}

pub fn match_ctor_impl(program_impl: &ItemImpl, sails_path: &Path) -> TokenStream {
    let (program_type_path, _, _) = shared::impl_type_refs(program_impl.self_ty.as_ref());
    let program_ctors = discover_program_ctors(program_impl);
    let ctor_branches =
        program_ctors
            .iter()
            .map(|(handler_route, (handler_fn, _, unwrap_result))| {
                ctor_branch_impl(
                    program_type_path,
                    handler_route,
                    handler_fn,
                    *unwrap_result,
                    sails_path,
                )
            });
    quote! {
        async fn match_ctor_solidity(ctor: &[u8], input: &[u8]) -> Option<#program_type_path> {
            #( #ctor_branches )*
            None
        }
    }
}

fn ctor_branch_impl(
    program_type_path: &TypePath,
    ctor_route: &str,
    ctor_fn: &ImplItemFn,
    unwrap_result: bool,
    sails_path: &Path,
) -> TokenStream {
    let handler_route_bytes = ctor_route.encode();
    let handler_func = Func::from(&ctor_fn.sig);
    let handler_ident = handler_func.ident();
    let handler_params = handler_func.params().iter().map(|item| {
        let param_ident = item.0;
        quote!(#param_ident)
    });
    let handler_params_comma = handler_func.params().iter().map(|item| {
        let param_ident = item.0;
        quote!(#param_ident,)
    });
    let handler_types = handler_func.params().iter().map(|item| {
        let param_type = item.1;
        quote!(#param_type,)
    });

    let await_token = handler_func.is_async().then(|| quote!(.await));
    let unwrap_token = unwrap_result.then(|| quote!(.unwrap()));

    quote! {
        if ctor == &[ #(#handler_route_bytes),* ] {
            let (#(#handler_params_comma)*) : (#(#handler_types)*) = #sails_path::alloy_sol_types::SolValue::abi_decode_params(input, false).expect("Failed to decode request");
            let program = #program_type_path :: #handler_ident (#(#handler_params),*) #await_token #unwrap_token;
            return Some(program);
        }
    }
}

fn service_signature(
    service_route: &str,
    service_fn: &ImplItemFn,
    unwrap_result: bool,
    sails_path: &Path,
) -> TokenStream {
    let service_route_bytes = service_route.encode();
    let service_name = service_route
        .with_boundaries(&[Boundary::DigitUpper, Boundary::UpperDigit])
        .to_case(Case::Snake);
    let service_type = shared::unwrap_result_type(&service_fn.sig, unwrap_result);

    quote! {
        (
            #service_name,
            &[ #(#service_route_bytes),* ] as &[u8],
            <#service_type as #sails_path::solidity::ServiceSignature>::METHODS,
        ),
    }
}
