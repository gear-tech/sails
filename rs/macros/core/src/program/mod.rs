use crate::{
    sails_paths,
    shared::{self, Func},
};
use args::ProgramArgs;
use parity_scale_codec::Encode;
use proc_macro2::{Span, TokenStream as TokenStream2};
use proc_macro_error::abort;
use quote::quote;
use std::collections::BTreeMap;
use syn::{
    parse_quote, spanned::Spanned, Ident, ImplItem, ImplItemFn, ItemImpl, Path, Receiver,
    ReturnType, Type, TypePath, Visibility,
};

mod args;

/// Static Span of Program `impl` block
static mut PROGRAM_SPAN: Option<Span> = None;

pub fn gprogram(args: TokenStream2, program_impl_tokens: TokenStream2) -> TokenStream2 {
    let program_impl = parse_gprogram_impl(program_impl_tokens);
    ensure_single_gprogram(&program_impl);
    let args = parse_args(args);
    gen_gprogram_impl(program_impl, args)
}

#[doc(hidden)]
pub fn __gprogram_internal(args: TokenStream2, program_impl_tokens: TokenStream2) -> TokenStream2 {
    let program_impl = parse_gprogram_impl(program_impl_tokens);
    let args = parse_args(args);
    gen_gprogram_impl(program_impl, args)
}

fn parse_args(args: TokenStream2) -> ProgramArgs {
    syn::parse2(args).unwrap_or_else(|err| {
        abort!(
            err.span(),
            "failed to parse `program` attribute arguments: {}",
            err
        )
    })
}

fn parse_gprogram_impl(program_impl_tokens: TokenStream2) -> ItemImpl {
    syn::parse2(program_impl_tokens).unwrap_or_else(|err| {
        abort!(
            err.span(),
            "`program` attribute can be applied to impls only: {}",
            err
        )
    })
}

fn ensure_single_gprogram(program_impl: &ItemImpl) {
    if unsafe { PROGRAM_SPAN }.is_some() {
        abort!(
            program_impl,
            "multiple `program` attributes are not allowed"
        )
    }
    unsafe { PROGRAM_SPAN = Some(program_impl.span()) };
}

fn gen_gprogram_impl(program_impl: ItemImpl, program_args: ProgramArgs) -> TokenStream2 {
    let sails_path = program_args.sails_path();
    let scale_codec_path = sails_paths::scale_codec_path(&sails_path);
    let scale_info_path = sails_paths::scale_info_path(&sails_path);

    let services_ctors = discover_services_ctors(&program_impl);

    let mut program_impl = program_impl.clone();

    let services_data = services_ctors
        .into_iter()
        .map(|(route, (ctor_fn, ctor_idx))| {
            let route_ident = Ident::new(
                &format!("__ROUTE_{}", route.to_ascii_uppercase()),
                Span::call_site(),
            );

            let route_static = {
                let ctor_route_bytes = route.encode();
                let ctor_route_len = ctor_route_bytes.len();
                quote!(
                    static #route_ident: [u8; #ctor_route_len] = [ #(#ctor_route_bytes),* ];
                )
            };

            let service_meta = {
                let service_type = shared::result_type(&ctor_fn.sig);
                quote!(
                    ( #route , #sails_path::meta::AnyServiceMeta::new::< #service_type >())
                )
            };

            wire_up_service_exposure(
                &mut program_impl,
                &route_ident,
                ctor_fn,
                ctor_idx,
                &sails_path,
            );

            (
                route_static,
                service_meta,
                ctor_fn.sig.ident.clone(),
                route_ident,
            )
        })
        .collect::<Vec<_>>();

    let program_ident = Ident::new("PROGRAM", Span::call_site());

    let handle_fn = generate_handle(
        &program_ident,
        services_data.iter().map(|item| (&item.3, &item.2)),
        program_args,
        &sails_path,
    );

    let services_meta = services_data.iter().map(|item| &item.1);

    let services_routes = services_data.iter().map(|item| &item.0);

    let (program_type_path, program_type_args) = shared::impl_type(&program_impl);
    let program_type_constraints = shared::impl_constraints(&program_impl);

    let (ctors_data, init_fn) = generate_init(
        &mut program_impl,
        &program_type_path,
        &program_ident,
        &sails_path,
    );

    let ctors_params_structs = ctors_data.clone().map(|item| item.2);

    let ctors_meta_variants = ctors_data.map(|item| {
        let ctor_route = Ident::new(&item.0, Span::call_site());
        let ctor_params_struct_ident = item.1;
        quote!(#ctor_route(#ctor_params_struct_ident))
    });

    quote!(
        #(#services_routes)*

        #program_impl

        impl #program_type_args #sails_path::meta::ProgramMeta for #program_type_path #program_type_constraints {
            fn constructors() -> #scale_info_path::MetaType {
                #scale_info_path::MetaType::new::<meta_in_program::ConstructorsMeta>()
            }

            fn services() -> impl Iterator<Item = (&'static str, #sails_path::meta::AnyServiceMeta)> {
                [
                    #(#services_meta),*
                ].into_iter()
            }
        }

        use #sails_path ::Decode as __ProgramDecode;
        use #sails_path ::TypeInfo as __ProgramTypeInfo;

        #(
            #[derive(__ProgramDecode, __ProgramTypeInfo)]
            #[codec(crate = #scale_codec_path )]
            #[scale_info(crate = #scale_info_path )]
            #[allow(dead_code)]
            #ctors_params_structs
        )*

        mod meta_in_program {
            use super::*;

            #[derive(__ProgramTypeInfo)]
            #[scale_info(crate = #scale_info_path )]
            pub enum ConstructorsMeta {
                #(#ctors_meta_variants),*
            }
        }

        #[cfg(target_arch = "wasm32")]
        pub mod wasm {
            use super::*;
            use #sails_path::{gstd, hex, prelude::*};

            static mut #program_ident: Option<#program_type_path> = None;

            #init_fn

            #handle_fn
        }
    )
}

fn wire_up_service_exposure(
    program_impl: &mut ItemImpl,
    route_ident: &Ident,
    ctor_fn: &ImplItemFn,
    ctor_idx: usize,
    sails_path: &Path,
) {
    let service_type = shared::result_type(&ctor_fn.sig);

    let mut original_service_ctor_fn = ctor_fn.clone();
    let original_service_ctor_fn_ident = Ident::new(
        &format!("__{}", original_service_ctor_fn.sig.ident),
        original_service_ctor_fn.sig.ident.span(),
    );
    original_service_ctor_fn.attrs.clear();
    original_service_ctor_fn.vis = Visibility::Inherited;
    original_service_ctor_fn.sig.ident = original_service_ctor_fn_ident.clone();
    program_impl
        .items
        .push(ImplItem::Fn(original_service_ctor_fn));

    let mut wrapping_service_ctor_fn = ctor_fn.clone();
    wrapping_service_ctor_fn.sig.output = parse_quote!(
        -> < #service_type as #sails_path::gstd::services::Service>::Exposure
    );
    wrapping_service_ctor_fn.block = parse_quote!({
        let service = self. #original_service_ctor_fn_ident ();
        let exposure = < #service_type as #sails_path::gstd::services::Service>::expose(
            service,
            #sails_path::gstd::msg::id().into(),
            #route_ident .as_ref(),
        );
        exposure
    });
    program_impl.items[ctor_idx] = ImplItem::Fn(wrapping_service_ctor_fn);
}

fn generate_init(
    program_impl: &mut ItemImpl,
    program_type_path: &TypePath,
    program_ident: &Ident,
    sails_path: &Path,
) -> (
    impl Iterator<Item = (String, Ident, TokenStream2)> + Clone,
    TokenStream2,
) {
    if discover_program_ctors(program_impl, program_type_path).is_empty() {
        program_impl.items.push(ImplItem::Fn(parse_quote!(
            pub fn default() -> Self {
                Self
            }
        )));
    }

    let program_ctors = discover_program_ctors(program_impl, program_type_path);

    let input_ident = Ident::new("input", Span::call_site());

    let mut invocation_dispatches = Vec::with_capacity(program_ctors.len());
    let mut invocation_params_structs = Vec::with_capacity(program_ctors.len());

    for (invocation_route, (program_ctor, ..)) in &program_ctors {
        let program_ctor = &program_ctor.sig;
        let handler = Func::from(program_ctor);

        let invocation_params_struct_ident =
            Ident::new(&format!("__{}Params", invocation_route), Span::call_site());

        invocation_dispatches.push({
            let invocation_route_bytes = invocation_route.encode();
            let invocation_route_len = invocation_route_bytes.len();
            let handler_ident = handler.ident();
            let handler_await = handler.is_async().then(|| quote!(.await));
            let handler_args = handler.params().iter().map(|item| {
                let param_ident = item.0;
                quote!(request.#param_ident)
            });

            quote!(
                if #input_ident.starts_with(& [ #(#invocation_route_bytes),* ]) {
                    static INVOCATION_ROUTE: [u8; #invocation_route_len] = [ #(#invocation_route_bytes),* ];
                    let request = #invocation_params_struct_ident::decode(&mut &#input_ident[#invocation_route_len..]).expect("Failed to decode request");
                    let program = #program_type_path :: #handler_ident (#(#handler_args),*) #handler_await;
                    (program, INVOCATION_ROUTE.as_ref())
                }
            )
        });

        invocation_params_structs.push({
            let invocation_params_struct_members = handler.params().iter().map(|item| {
                let param_ident = item.0;
                let param_type = item.1;
                quote!(#param_ident: #param_type)
            });

            (
                invocation_route.clone(),
                invocation_params_struct_ident.clone(),
                quote!(
                    struct #invocation_params_struct_ident {
                        #(#invocation_params_struct_members),*
                    }
                ),
            )
        });
    }

    invocation_dispatches.push(shared::generate_unexpected_input_panic(
        &input_ident,
        "Unexpected ctor",
        sails_path,
    ));

    let init = quote!(
        #[gstd::async_init]
        async fn init() {
            #sails_path::gstd::events::__enable_events();
            let mut #input_ident: &[u8] = &gstd::msg::load_bytes().expect("Failed to read input");
            let (program, invocation_route) = #(#invocation_dispatches)else*;
            unsafe {
                #program_ident = Some(program);
            }
            gstd::msg::reply_bytes(invocation_route, 0).expect("Failed to send output");
        }
    );

    (invocation_params_structs.into_iter(), init)
}

fn generate_handle<'a>(
    program_ident: &'a Ident,
    service_ctors: impl Iterator<Item = (&'a Ident, &'a Ident)>,
    program_args: ProgramArgs,
    sails_path: &Path,
) -> TokenStream2 {
    let input_ident = Ident::new("input", Span::call_site());

    let mut invocation_dispatches = Vec::new();

    for (service_route_ident, service_ctor_ident) in service_ctors {
        invocation_dispatches.push({
            quote!(
                if #input_ident.starts_with(& #service_route_ident) {
                    let program_ref = unsafe { #program_ident.as_ref() }.expect("Program not initialized");
                    let mut service = program_ref.#service_ctor_ident();
                    let output = service.handle(&#input_ident[#service_route_ident .len()..]).await;
                    [#service_route_ident .as_ref(), &output].concat()
                }
            )
        });
    }

    invocation_dispatches.push(shared::generate_unexpected_input_panic(
        &input_ident,
        "Unexpected service",
        sails_path,
    ));

    let mut args = Vec::with_capacity(2);
    if let Some(handle_reply) = program_args.handle_reply() {
        args.push(quote!(handle_reply = #handle_reply));
    }
    if let Some(handle_signal) = program_args.handle_signal() {
        args.push(quote!(handle_signal = #handle_signal));
    }
    let async_main_args = if args.is_empty() {
        quote!()
    } else {
        quote!((#(#args),*))
    };

    quote!(
        #[gstd::async_main #async_main_args]
        async fn main() {
            let mut #input_ident: &[u8] = &gstd::msg::load_bytes().expect("Failed to read input");
            let output: Vec<u8> = #(#invocation_dispatches)else*;
            gstd::msg::reply_bytes(output, 0).expect("Failed to send output");
        }
    )
}

fn discover_program_ctors<'a>(
    program_impl: &'a ItemImpl,
    program_type_path: &'a TypePath,
) -> BTreeMap<String, (&'a ImplItemFn, usize)> {
    let self_type_path = syn::parse_str::<TypePath>("Self").unwrap();
    shared::discover_invocation_targets(program_impl, |fn_item| {
        if matches!(fn_item.vis, Visibility::Public(_)) && fn_item.sig.receiver().is_none() {
            if let ReturnType::Type(_, output_type) = &fn_item.sig.output {
                if let Type::Path(output_type_path) = output_type.as_ref() {
                    if output_type_path == &self_type_path || output_type_path == program_type_path
                    {
                        return true;
                    }
                }
            }
        }
        false
    })
}

fn discover_services_ctors(program_impl: &ItemImpl) -> BTreeMap<String, (&ImplItemFn, usize)> {
    shared::discover_invocation_targets(program_impl, |fn_item| {
        matches!(fn_item.vis, Visibility::Public(_))
            && matches!(
                fn_item.sig.receiver(),
                Some(Receiver {
                    mutability: None,
                    reference: Some(_),
                    ..
                })
            )
            && fn_item.sig.inputs.len() == 1
            && !matches!(fn_item.sig.output, ReturnType::Default)
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use quote::quote;

    #[test]
    fn gprogram_discovers_public_associated_functions_returning_self_or_the_type_as_ctors() {
        let program_impl = syn::parse2(quote!(
            impl MyProgram {
                fn non_public_associated_func_returning_self() -> Self {}
                fn non_public_associated_func_returning_type() -> MyProgram {}
                fn non_public_associated_func_returning_smth() -> u32 {}
                pub fn public_associated_func_returning_self() -> Self {}
                pub fn public_associated_func_returning_type() -> MyProgram {}
                pub fn public_associated_func_returning_smth() -> u32 {}
                fn non_public_method_returning_self(&self) -> Self {}
                fn non_public_method_returning_type(&self) -> MyProgram {}
                fn non_public_method_returning_smth(&self) -> u32 {}
                pub fn public_method_returning_self(&self) -> Self {}
                pub fn public_method_returning_type(&self) -> MyProgram {}
                pub fn public_method_returning_smth(&self) -> u32 {}
            }
        ))
        .unwrap();
        let (program_type_path, ..) = shared::impl_type(&program_impl);

        let discovered_ctors = discover_program_ctors(&program_impl, &program_type_path)
            .iter()
            .map(|(.., (ctor_fn, ..))| ctor_fn.sig.ident.to_string())
            .collect::<Vec<_>>();

        assert_eq!(discovered_ctors.len(), 2);
        assert!(discovered_ctors.contains(&String::from("public_associated_func_returning_self")));
        assert!(discovered_ctors.contains(&String::from("public_associated_func_returning_type")));
    }

    #[test]
    fn gprogram_discovers_public_methods_with_self_ref_only_and_some_return_as_service_funcs() {
        let program_impl = syn::parse2(quote!(
            impl MyProgram {
                fn non_public_associated_func_returning_smth() -> u32 {}
                fn non_public_associated_func_returning_unit() {}
                pub fn public_associated_func_returning_smth() -> MyProgram {}
                pub fn public_associated_func_returning_unit() {}
                fn non_public_method_returning_smth(&self) -> u32 {}
                fn non_public_method_returning_unit(&self) {}
                pub fn public_method_returning_smth(&self) -> u32 {}
                pub fn public_method_returning_smth_with_other_params(&self, p1: u32) -> u32 {}
                pub fn public_methos_returning_smth_and_consuming_self(self) -> u32 {}
            }
        ))
        .unwrap();

        let discovered_services = discover_services_ctors(&program_impl)
            .iter()
            .map(|(_, (fn_impl, _))| fn_impl.sig.ident.to_string())
            .collect::<Vec<_>>();

        assert_eq!(discovered_services.len(), 1);
        assert!(discovered_services.contains(&String::from("public_method_returning_smth")));
    }
}
