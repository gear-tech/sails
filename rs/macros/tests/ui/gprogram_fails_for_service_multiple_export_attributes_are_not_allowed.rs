use sails_macros::program;

struct MyService;

struct MyProgram;

#[program]
impl MyProgram {
    #[export(route = "svc1")]
    #[export(route = "svc2")]
    pub fn service1(&self) -> MyService {
        MyService
    }
}

#[tokio::main]
async fn main() {}
