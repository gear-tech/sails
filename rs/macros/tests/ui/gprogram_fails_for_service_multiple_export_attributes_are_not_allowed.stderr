error: multiple `export` attributes are not allowed for the same method
  --> tests/ui/gprogram_fails_for_service_multiple_export_attributes_are_not_allowed.rs:10:7
   |
10 |     #[export(route = "svc2")]
   |       ^^^^^^
